parser/prism/Translator.cc

```diff
@@ -792,22 +820,63 @@ unique_ptr<parser::Node> Translator::translate(pm_node_t *node, bool preserveCon
 
             int numPosArgs = args.size() - (hasKwargsHash ? 1 : 0);
 
+            if (prismBlock != nullptr && PM_NODE_TYPE_P(prismBlock, PM_BLOCK_ARGUMENT_NODE) && !blockPassArgIsSymbol) {
+                // Special handling for non-Symbol block pass args, like `a.map(&block)`
+                // Symbol procs like `a.map(:to_s)` are rewritten into literal block arguments,
+                // and handled separately below.
+
+                // Desugar a call without a splat, and any other expression as a block pass argument.
+                // E.g. `a.each(&block)`
+
+                ast::Send::ARGS_store magicSendArgs;
+                magicSendArgs.reserve(3 + args.size());
+                magicSendArgs.emplace_back(move(receiverExpr));
+                magicSendArgs.emplace_back(move(methodName));
+                magicSendArgs.emplace_back(move(blockPassArg));
+
+                numPosArgs += 3;
+
+                for (auto &arg : args) {
+                    magicSendArgs.emplace_back(arg->takeDesugaredExpr());
+                }
+
+                auto sendExpr = MK::Send(loc, MK::Magic(loc), core::Names::callWithBlockPass(), messageLoc, numPosArgs,
+                                         move(magicSendArgs), flags);
+
+                return make_node_with_expr<parser::Send>(move(sendExpr), loc, move(receiver), name, messageLoc,
+                                                         move(args));
+            }
+
             ast::Send::ARGS_store sendArgs{};
-            sendArgs.reserve(args.size());
+            sendArgs.reserve(args.size()); // TODO: reserve size for block, if needed.
             for (auto &arg : args) {
                 sendArgs.emplace_back(arg->takeDesugaredExpr());
             }
 
             if (prismBlock != nullptr) {
-                if (PM_NODE_TYPE_P(prismBlock, PM_BLOCK_NODE)) {
-                    auto blockBodyExpr = blockBody == nullptr ? MK::EmptyTree() : blockBody->takeDesugaredExpr();
-                    auto blockExpr = MK::Block(location, move(blockBodyExpr), move(blockParamsStore));
+                if (PM_NODE_TYPE_P(prismBlock, PM_BLOCK_NODE) || blockPassArgIsSymbol) {
+                    // A literal block arg (like `foo { ... }` or `foo do ... end`),
+                    // or a Symbol proc like `&:b` (which we'll desugar into a literal block)
+
+                    ast::ExpressionPtr blockExpr;
+                    if (blockPassArgIsSymbol) {
+                        ENFORCE(PM_NODE_TYPE_P(prismBlock, PM_BLOCK_ARGUMENT_NODE));
+                        auto *bp = down_cast<pm_block_argument_node>(prismBlock);
+                        ENFORCE(PM_NODE_TYPE_P(bp->expression, PM_SYMBOL_NODE));
+                        ENFORCE(false, "TODO: Implement symbol block pass args here");
```

Comment:
Symbol procs will always hit the "fallback" path for now, because of the `supportedBlock = false; // TODO: Implement symbol procs` above.